### PR TITLE
Add some clarifying comments and refactor a portion of the `Keyring` class for readability

### PR DESCRIPTION
**Pull request by anoadragon453**
This caused me a great deal of confusion both back when writing https://github.com/matrix-org/synapse/issues/12767 and now. I've added some clarifying comments and refactoring some of the logic in the `Keyring` class.

Logic should not have changed, but please double-check. Closes https://github.com/matrix-org/synapse/issues/12767.

### CODE REVIEW

File: synapse/crypto/keyring.py [hunk 1]
```diff
@@ -279,6 +282,11 @@ async def process_request(self, verify_request: VerifyJsonRequest) -> None:
 
         key_ids_to_find = set(verify_request.key_ids) - found_keys.keys()
         if key_ids_to_find:
+            # We're still missing some keys. Consult each of our `KeyFetcher` instances
+            # (stored in `self._key_fetchers`) to try and find them.
+            # Key fetch attempts are queued via `self._server_queue` below, and carried
+            # out in `self._inner_fetch_key_requests`.
```
Comment (clokep) [inline synapse/crypto/keyring.py:288]: Most of this feels like it should be in the docstring to `add_to_queue`? (or the `_inner_fetch_key_request` or `Keyring` or `_key_fetchers` or something?)

I think putting it here will bury it.

Comment (anoadragon453) [reply synapse/crypto/keyring.py:288]: I thought about that, though in practice at this bit in the codebase I saw:

```
found_keys_by_server = await self._server_queue.add_to_queue(
    key_request, key=verify_request.server_name
)
```

and assumed that keys were being fetched from the remote server. Having this comment here would have helped. We could potentially rename `_server_queue` to `_fetch_keys_queue` as well/instead.

Comment (clokep) [reply synapse/crypto/keyring.py:288]: I think renaming server queue is likely the way to go? And maybe move the comment to where it is declared?

File: synapse/crypto/keyring.py [hunk 2]
```diff
@@ -279,6 +282,11 @@ async def process_request(self, verify_request: VerifyJsonRequest) -> None:
 
         key_ids_to_find = set(verify_request.key_ids) - found_keys.keys()
         if key_ids_to_find:
+            # We're still missing some keys. Consult each of our `KeyFetcher` instances
```
Comment (dklimpel) [inline synapse/crypto/keyring.py:285]: Is it possible to set / configure which Synapse / worker instance is a `KeyFetcher`?

Comment (dklimpel) [reply synapse/crypto/keyring.py:285]: Is this the answer #14805 ?

Comment (anoadragon453) [reply synapse/crypto/keyring.py:285]: Yep!

**Comment by anoadragon453** [comment]
Have renamed the queue and left a trail of docstring breadcrumbs. I think this is indeed much better, thank you!

File: synapse/crypto/keyring.py [hunk 3]
```diff
@@ -420,26 +449,22 @@ async def _inner_fetch_key_request(
...
 
-                # If we already have a result for the given key ID we keep the
+                # If we already have a result for the given key ID, we keep the
                 # one with the highest `valid_until_ts`.
                 existing_key = found_keys.get(key_id)
-                if existing_key:
-                    if key.valid_until_ts <= existing_key.valid_until_ts:
-                        continue
+                if existing_key and existing_key.valid_until_ts >= key.valid_until_ts:
+                    continue
```
Comment (clokep) [inline synapse/crypto/keyring.py:456]: By going from `a <= b` to `b >= a` I think we've swapped what `a == b` does.

Comment (anoadragon453) [reply synapse/crypto/keyring.py]: Hrm, indeed. That serves me right for trying to eyeball boolean logic on the fly. Have fixed in [357eef8](https://github.com/matrix-org/synapse/pull/14804/commits/357eef8982cedf2047db338112b60c8be3f94101).

**Review by clokep — APPROVED** [review]

**Pull request merged**